spark3/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedure.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.procedures;
+
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.iceberg.catalog.Procedure;
+
+public enum SparkProcedure {
+  ROLLBACK_TO_SNAPSHOT(RollbackToSnapshotProcedure::new);
+
+  private final ProcedureBuilder builder;
+
+  SparkProcedure(ProcedureBuilder builder) {
+    this.builder = builder;
+  }
+
+  public Procedure build(TableCatalog catalog) {
+    return builder.build(catalog);
+  }
+
+  private interface ProcedureBuilder {
+    Procedure build(TableCatalog catalog);
```

Comment:
It seems strange that the builder passes `catalog` to build, rather than `withTableCatalog`.

Reply:
Do you mean the overall approach or the method name?

Reply:
I'm assuming that there will be more options that will be passed to the builder, but there is just TableCatalog for now. I would expect this to support cases that don't need a TableCatalog, so it doesn't seem like something that we should tie to the build method. Instead, I'd use a `withTableCatalog` method and always build the procedure with `build()`

Reply:
Let me rework this part.

Reply:
Updated. Let me know what you think, @rdblue.